### PR TITLE
Sparse state preparation via Dense + Permute

### PR DESCRIPTION
**Pull request by anurudhp**

### CODE REVIEW

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    This bloq prepares a $d$-sparse state on $n$ qubits:
+
+        $$
+            \ket{\psi} = \sum_{x \in S} \alpha_x \ket{x}
+        $$
+
+    where $S \subseteq [N]$ such that $|S| = d$.
+
+    To achieve this, it first prepares a dense state on $\ceil{\log d}$ qubits, then
+    permutes the basis s.t. $i \mapsto x_i$, where $x_i$ is the $i$-th element in S.
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py:50]: added

File: qualtran/conftest.py [hunk 2]
```diff
@@ -111,6 +111,7 @@ def assert_bloq_example_serializes_for_pytest(bloq_ex: BloqExample):
         'explicit_matrix_block_encoding',  # cannot serialize AutoPartition
         'symmetric_banded_matrix_block_encoding',  # cannot serialize AutoPartition
         'chebyshev_poly_even',
+        'sparse_state_prep_via_rotations',
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/conftest.py]: added

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+from qualtran.bloqs.basic_gates import IntState
+from qualtran.bloqs.rotations import PhaseGradientState
+from qualtran.bloqs.state_preparation.sparse_state_preparation_via_rotations import (
+    _sparse_state_prep_via_rotations,
+    SparseStatePreparationViaRotations,
+)
+
+
+def test_examples(bloq_autotester):
+    bloq_autotester(_sparse_state_prep_via_rotations)
```
Comment (mpharrigan) [inline qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py:28]: this is triggering scipy errors on my machine. Is there a particular version of scipy that is needed?

Comment (mpharrigan) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py:28]:
```
bloq_autotester = <function assert_bloq_example_make_for_pytest at 0x11a417e20>

    def test_examples(bloq_autotester):
>       bloq_autotester(_sparse_state_prep_via_rotations)

sparse_state_preparation_via_rotations_test.py:28: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../conftest.py:39: in assert_bloq_example_make_for_pytest
    qlt_testing.assert_bloq_example_make(bloq_ex)
../../testing.py:368: in assert_bloq_example_make
    bloq = bloq_ex.make()
../../_infra/bloq_example.py:53: in make
    return self._func()
sparse_state_preparation_via_rotations.py:184: in _sparse_state_prep_via_rotations
    sparse_state_prep_via_rotations = SparseStatePreparationViaRotations.from_sparse_array(
sparse_state_preparation_via_rotations.py:101: in from_sparse_array
    sparse_coeffs = scipy.sparse.dok_array(coeffs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[TypeError("'NoneType' object is not iterable") raised in repr()] dok_array object at 0x17aab0690>
arg1 = array([0.70914953, 0.        , 0.        , 0.        , 0.46943701,
       0.        , 0.2297245 , 0.        , 0.        , 0.32960471,
       0.        , 0.        , 0.33959273, 0.        , 0.        ])
shape = None, dtype = None, copy = False

    def __init__(self, arg1, shape=None, dtype=None, copy=False):
        _spbase.__init__(self)
        self._dict = {}
    
        self.dtype = getdtype(dtype, default=float)
        if isinstance(arg1, tuple) and isshape(arg1):  # (M,N)
            M, N = arg1
            self._shape = check_shape((M, N))
        elif issparse(arg1):  # Sparse ctor
            if arg1.format == self.format and copy:
                arg1 = arg1.copy()
            else:
                arg1 = arg1.todok()
    
            if dtype is not None:
                arg1 = arg1.astype(dtype, copy=False)
    
            self._dict.update(arg1)
            self._shape = check_shape(arg1.shape)
            self.dtype = arg1.dtype
        else:  # Dense ctor
            try:
                arg1 = np.asarray(arg1)
            except Exception as e:
                raise TypeError('Invalid input format.') from e
    
            if len(arg1.shape) != 2:
>               raise TypeError('Expected rank <=2 dense array or matrix.')
E               TypeError: Expected rank <=2 dense array or matrix.
```

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        return bit_length(slen(self.sparse_indices) - 1)
+
+    def is_symbolic(self):
+        return is_symbolic(self.sparse_indices, self.nonzero_coeffs, self.N)
+
+    @classmethod
+    def from_coefficient_map(
+        cls, N: SymbolicInt, coeff_map: dict[int, complex], phase_bitsize: SymbolicInt
+    ):
+        """Factory"""
```
Comment (mpharrigan) [inline qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py:88]: write a real docstring

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        return cls(tuple(coeff_map.keys()), tuple(coeff_map.values()), N, phase_bitsize)
+
+    @classmethod
+    def from_sparse_array(
+        cls,
+        coeffs: Union[Sequence[complex], NDArray[np.complex_], 'scipy.sparse.sparray'],
+        phase_bitsize: SymbolicInt,
+    ):
+        """Factory"""
```
Comment (mpharrigan) [inline qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py:97]: docstring

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        sparse_coeffs = scipy.sparse.dok_array(coeffs)
+        sparse_coeffs_as_dict = dict(sparse_coeffs.items())
+
+        return cls.from_coefficient_map(N, sparse_coeffs_as_dict, phase_bitsize)
+
+    @classmethod
+    def from_n_coeffs(
+        cls, n_coeffs: SymbolicInt, n_nonzero_coeffs: SymbolicInt, phase_bitsize: SymbolicInt
+    ):
+        """Factory"""
```
Comment (mpharrigan) [inline qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py:110]: docstring

File: qualtran/linalg/permutation.py [hunk 7]
```diff
@@ -47,7 +47,7 @@ def decompose_permutation_map_into_cycles(permutation_map: dict[int, int]) -> It
 
         # compute the cycle starting at `i`
         cycle = []
-        while True:
+        while i not in seen:
```
Comment (mpharrigan) [inline qualtran/linalg/permutation.py:50]: add a test that would have failed?

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py (hunk 3 above)
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py:28]: I am using the one pinned in the dependencies, 1.14.0. Which version are you using?

Comment (mpharrigan) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py:28]: 1.12.0

Comment (mpharrigan) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py:28]: which was released jan, 2024.

We should generally aim to be compatible with each version permissible by `runtime.txt`, so if you need a particularly new feature: provide a minimum version. Unless there's a really really important new feature in the past 8 months of scipy; we should probably re-work the example to be more compatible

File: qualtran/linalg/permutation.py (hunk 7 above)
Comment (anurudhp) [reply qualtran/linalg/permutation.py:50]: added in https://github.com/quantumlib/Qualtran/pull/1307

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py (hunk 4 above)
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py:88]: #1306

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py (hunk 5 above)
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py:97]: https://github.com/quantumlib/Qualtran/pull/1306

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py (hunk 6 above)
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations.py:110]: https://github.com/quantumlib/Qualtran/pull/1306

File: qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py (hunk 3 above)
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/sparse_state_preparation_via_rotations_test.py:28]: #1308

**Pull request merged**